src/analysis/retail/shaman/elemental/modules/features/SpenderWindow.tsx

```diff
@@ -0,0 +1,295 @@
+];
+
+const GOOD_SOP_SPENDERS = [SPELLS.LIGHTNING_BOLT.id, TALENTS.CHAIN_LIGHTNING_TALENT.id];
+
+const PERFECT_WINDOWS_THRESHOLD: GTEThreshold = {
+  type: 'gte',
+  perfect: 0.9,
+  good: 0.8,
+  ok: 0.7,
+};
+const WRONG_SOP_THRESHOLD: LTEThreshold = {
+  type: 'lte',
+  perfect: 0,
+  good: 0.1,
+  ok: 0.15,
+};
+const MISSING_ELSHOCKS_THRESHOLD: LTEThreshold = {
+  type: 'lte',
+  perfect: 0,
+  good: 0.2,
+  ok: 0.3,
+};
+const MISSING_MOTE_THRESHOLD: LTEThreshold = {
+  type: 'lte',
+  perfect: 0,
+  good: 0.2,
+  ok: 0.3,
+};
+const MISSING_FLAMESHOCK_THRESHOLD: LTEThreshold = {
+  type: 'lte',
+  perfect: 0,
+  good: 0.2,
+  ok: 0.3,
+};
+
+class SpenderWindow extends Analyzer {
+  static dependencies = {
+    enemies: Enemies,
+  };
+
+  enabled: boolean;
+
+  protected enemies!: Enemies;
+
+  activeSpenderWindow: ActiveSpenderWindow | null;
+  spenderWindows: FinishedSpenderWindow[];
+
+  private stSpender: Talent = TALENTS.EARTH_SHOCK_TALENT;
+
+  constructor(options: Options) {
+    super(options);
+
+    this.activeSpenderWindow = null;
+    this.spenderWindows = [];
+
+    if (this.selectedCombatant.hasTalent(TALENTS.ELEMENTAL_BLAST_TALENT)) {
+      this.stSpender = TALENTS.ELEMENTAL_BLAST_TALENT;
+    }
+
+    this.enabled = this.selectedCombatant.hasTalent(TALENTS.SURGE_OF_POWER_TALENT);
+
+    /* There is no point in tracking this at all if the player doesn't have SoP */
+    if (!this.enabled) {
+      return;
+    }
+
+    this.addEventListener(Events.cast.by(SELECTED_PLAYER).spell(this.stSpender), this.onSpender);
+    this.addEventListener(Events.cast.by(SELECTED_PLAYER).spell(SOP_SPENDERS), this.onSopSpender);
+  }
+
+  onSpender(event: CastEvent) {
+    if (this.activeSpenderWindow) {
+      return;
+    }
+
+    this.activeSpenderWindow = {
+      timestamp: event.timestamp,
+      motePresent: this.selectedCombatant.hasBuff(
+        SPELLS.MASTER_OF_THE_ELEMENTS_BUFF.id,
+        null,
+        ON_CAST_BUFF_REMOVAL_GRACE_MS,
+      ),
+      flameshockPresent:
+        this.enemies.getEntity(event)?.hasBuff(SPELLS.FLAME_SHOCK.id, event.timestamp) || false,
+    };
+  }
+
+  onSopSpender(event: CastEvent) {
+    if (!this.activeSpenderWindow) {
+      return;
+    }
+    if (
+      !this.selectedCombatant.hasBuff(
+        SPELLS.SURGE_OF_POWER_BUFF.id,
+        null,
+        ON_CAST_BUFF_REMOVAL_GRACE_MS,
+      )
+    ) {
+      return;
+    }
+
+    const elshocksPresent =
+      this.enemies
+        .getEntity(event)
+        ?.hasBuff(SPELLS.ELECTRIFIED_SHOCKS_DEBUFF.id, event.timestamp) || false;
+
+    this.spenderWindows.push({
+      ...this.activeSpenderWindow,
+      elshocksPresent,
+      sopUse: event,
+    });
+    this.activeSpenderWindow = null;
+  }
+
+  private spenderWindowsRow(
+    subWindow: FinishedSpenderWindow[],
+    header: JSX.Element,
+    performanceThresholds: LTEThreshold | GTEThreshold,
+    windowTimestampCallable: (windows: FinishedSpenderWindow[]) => JSX.Element[] = (windows) =>
```

Comment:
[Consider] If this is never used, why have it here as a default argument?  I guessing this was left over for something.  Might be worth a comment.

Reply:
It is used in the inefficient SoP row to add the spell that was actually used to the tooltip.